clients/slo/slo.go

```diff
@@ -0,0 +1,213 @@
+// @license
+// Copyright 2025 Dynatrace LLC
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slo
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/clients/slo"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/rest"
+)
+
+func NewClient(client *rest.Client) *Client {
+	c := &Client{
+		client: slo.NewClient(client),
+	}
+	return c
+}
+
+type Client struct {
+	client client
+}
+
+//go:generate mockgen -source slo.go -package=slo -destination=client_mock.go
+type client interface {
+	List(ctx context.Context, ro rest.RequestOptions) (*http.Response, error)
+	Get(ctx context.Context, id string, ro rest.RequestOptions) (*http.Response, error)
+	Create(ctx context.Context, data []byte, ro rest.RequestOptions) (*http.Response, error)
+	Update(ctx context.Context, id string, optimisticLockingVersion string, data []byte, ro rest.RequestOptions) (*http.Response, error)
+	Delete(ctx context.Context, id string, ro rest.RequestOptions) (*http.Response, error)
+}
+
+var _ client = (*slo.Client)(nil)
+
+const errMsgWithId = "failed to %s slo resource with id %s: %w"
+
+// List gets a complete set of complete configuration for all available SLOs
+func (c Client) List(ctx context.Context) (api.PagedListResponse, error) {
+	var retVal api.PagedListResponse
+
+	for haveNextPage, nextPageKey := true, ""; haveNextPage; {
+		response, err := c.client.List(ctx, rest.RequestOptions{
+			CustomShouldRetryFunc: rest.RetryIfTooManyRequests,
+			QueryParams:           url.Values{"page-key": {nextPageKey}}})
+		defer closeBody(response)
```

Comment:
This (inner loop) should be extracted in a function, as this is a minor resource leak (body stays open until all resources have been downloaded)

Reply:
done